runtime/common/marshaling/index.ts

```diff
@@ -1,106 +1,154 @@
+      pathPrefix.push(index.toString());
+      const {val: itemVal, hasModifications} = fixUncopyableTypes(item, pathPrefix, postTransforms, depth + 1);
+      if (hasModifications) {
+        someItemHadModifications = true;
+      }
+      maybeModifiedArray.push(itemVal);
+      pathPrefix.pop();
+    });
+    if (someItemHadModifications) {
+      return {val: maybeModifiedArray, hasModifications: true};
     }
+  }
 
-    // if val has a constructor that isn't recognized by customMarshalers, it will be
-    // converted into a plain object.
-    const processedValue: {[key: string]: any} = {};
-    for (const key of Object.getOwnPropertyNames(objectValue)) {
-      processedValue[key] = processValue(objectValue[key], depth + 1);
+  if (typeof val === 'object') {
+    const maybeModifiedObject: any = {};
+    let hadModifications = false;
+    for (const key of Object.getOwnPropertyNames(val)) {
+      pathPrefix.push(key);
+      const {val: objVal, hasModifications: subValHasModifications} = fixUncopyableTypes(
+        val[key],
+        pathPrefix,
+        postTransforms,
+        depth + 1,
+      );
+      maybeModifiedObject[key] = objVal;
+      pathPrefix.pop();
+      if (subValHasModifications) {
+        hadModifications = true;
+      }
+    }
+    if (hadModifications) {
+      return {val: maybeModifiedObject, hasModifications: true};
     }
-
-    return processedValue;
   }
 
-  return serializedValue;
+  return {val, hasModifications: false};
+}
+
+function isMarshaledValue(val: any): boolean {
+  return typeof val === 'object' && MarshalingInjectedKeys.CodaMarshaler in val;
 }
 
-export function marshalValue(val: any): string | undefined {
-  // Instead of passing a replacer to `JSON.stringify`, we chose to preprocess the value before
-  // passing it to `JSON.stringify`. The reason is that `JSON.stringify` may call the object toJSON
-  // method before calling the replacer. In many cases, that means the replacer can't tell if the
-  // input has been processed by toJSON or not. For example, Date.prototype.toJSON simply returns
-  // a date string. The replacer can't identify if the initial value is a Date or a date string.
-  //
-  // processValue is trying to mimic the object processing of JSON but the behavior may not be
-  // identical. It will only serve the purpose of our internal marshaling use case.
-  return JSON.stringify(processValue(val));
+export function marshalValue(val: any): any {
+  const postTransforms: object[] = [];
+  const {val: encodedVal} = fixUncopyableTypes(val, [], postTransforms, 0);
+  const result = {
+    encoded: encodedVal,
+    postTransforms,
+    [MarshalingInjectedKeys.CodaMarshaler]: CodaMarshalerType.Object,
+  };
+  return result;
 }
 
-export function unmarshalValue(marshaledValue: string | undefined): any {
-  if (marshaledValue === undefined) {
-    return marshaledValue;
+function applyTransform(input: any, path: string[], fn: (encoded: any) => any): any {
```

Comment:
I think this approach works. but I slightly lean towards an in-place marshaling where you can easily apply the same traverse function with a marshaler / unmarshaler function. 

but I do see the postTransforms could be more efficient in most cases.

Reply:
let me add another note on this comment. I think the `postTransforms` is more efficient here while I was considering complicated cases like nested marshaling (e.g. Error has a property of Buffer). When the non-copyable types get more complicated, in-place marshaling would be more adaptive while you need to carefully control the `postTransforms` ordering and paths (in case unmarshaling changes the hierarchy of a nested object).

Reply:
I don't think the old implementation recursively supported buffers within errors either

Reply:
Hypothetically if we want to supported buffers in errors, we could recursively call marsalValue and the post transforms would be embedded deep within the output instead of merged into the top level

Reply:
I just verified that buffer in error works in the old form. It's a real bug we found before and had to apply a fix. would be a bit surprised if it no longer works. 

You may be able to make it work in this PR by not directly returning when it detects an error, and keep serializing its sub fields. I was just worried if the path in `postTransforms` is going to be precise.

Reply:
I think I can actually fix all the issues with errors, but still trying to get the code working. Looks like `v8.serialize()` and `v8.deserialize()` exist and are compatible with structure clone, so one thing I plan to do is encode errors as `new Error(Buffer.from(v8.serialize(marshalValue(...)), "base64"))`

Also planning to make errors recursively marshal any spare arguments so errors-within-errors work properly

Reply:
I am not sure if I follow what you plan to do. 

on the other hand, it looks like v8.serialize/deserialize doesn't seem to work for custom errors. 

```
> class NewError extends Error {}
undefined
> v8.deserialize(v8.serialize(new NewError()))
Error
    at REPL14:1:29
    at Script.runInThisContext (node:vm:129:12)
    at REPLServer.defaultEval (node:repl:562:29)
    at bound (node:domain:421:15)
    at REPLServer.runBound [as eval] (node:domain:432:12)
    at REPLServer.onLine (node:repl:889:10)
    at REPLServer.emit (node:events:402:35)
    at REPLServer.emit (node:domain:475:12)
    at REPLServer.Interface._onLine (node:readline:487:10)
    at REPLServer.Interface._line (node:readline:864:8)
> v8.deserialize(v8.serialize(new NewError())) instanceof NewError
false
```

It looks like the same behavior of how isolated-vm passes error objects around. we need a special error marshaling to keep all non-standard error information.

Reply:
`v8.serialize` isn't for simplifying the logic, it's for fixing the problem with `wrapError` since isolated-vm errors only seem to be able to bubble up a string and this new `marshalValue` implementation wasn't returning a string

If you look at the unit tests currently in place, they show that `wrapError`/`unwrapError` are losing type information. I should be able to fix that with `serialize`/`deserialize`

Reply:
oh you are right. just noticed that.